app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt

```diff
@@ -41,29 +40,32 @@ object MessageController {
     //endregion
 
     //region Edit data
-    fun update(apiMessages: List<Message>, thread: Thread) {
-
-        // Get current data
-        Log.d(RealmDatabase.TAG, "Messages: Get current data")
-        val realmMessages = thread.messages
+    fun update(realmMessages: List<Message>, apiMessages: List<Message>) {
+        RealmDatabase.mailboxContent.writeBlocking {
 
-        // Get outdated data
-        Log.d(RealmDatabase.TAG, "Messages: Get outdated data")
-        // val deletableMessages = MailboxContentController.getDeletableMessages(messagesFromApi)
-        val deletableMessages = realmMessages.filter { realmMessage ->
-            apiMessages.none { apiMessage -> apiMessage.uid == realmMessage.uid }
-        }
+            // Get outdated data
+            Log.d(RealmDatabase.TAG, "Messages: Get outdated data")
+            val outdatedMessages = getOutdatedMessages(realmMessages, apiMessages)
 
-        RealmDatabase.mailboxContent.writeBlocking {
             // Save new data
             Log.d(RealmDatabase.TAG, "Messages: Save new data")
-            apiMessages.forEach { apiMessage ->
-                if (!apiMessage.isManaged()) copyToRealm(apiMessage, UpdatePolicy.ALL)
-            }
+            insertNewData(apiMessages)
 
             // Delete outdated data
             Log.d(RealmDatabase.TAG, "Messages: Delete outdated data")
-            deleteMessages(deletableMessages)
+            deleteMessages(outdatedMessages)
+        }
+    }
+
+    private fun getOutdatedMessages(realmMessages: List<Message>, apiMessages: List<Message>): List<Message> {
+        return realmMessages.filter { realmMessage ->
+            apiMessages.none { apiMessage -> apiMessage.uid == realmMessage.uid }
+        }
+    }
```

Comment:
Simplify this with a realm query

Reply:
Blocked by https://github.com/realm/realm-kotlin/issues/591